cookbook/case_studies/ml_training/nlp_processing/word2vec_and_lda.py

```diff
@@ -0,0 +1,342 @@
+# ===================
+#
+# We compute Word Mover’s Distance (WMD) using the trained embeddings of words.
+# This enables us to assess the distance between two documents in a meaningful way even when they have
+# no words in common.
+# WMD outputs a large value for two completely unrelated sentences and small value for two closely related
+# sentences.
+# Since we chose two similar sentences for comparison, the word movers distance
+# should be small. You can try altering either ``SENTENCE_A`` or ``SENTENCE_B`` variables to be dissimilar
+# to the other sentence, and check if the value computed is larger.
+@task(cache_version="1.0", cache=True, limits=Resources(mem="600Mi"))
+def word_movers_distance(model_ser: FlyteFile[MODELSER_NLP]):
+    sentences = [SENTENCE_A, SENTENCE_B]
+    results = []
+    for i in sentences:
+        result = [w for w in utils.tokenize(i) if w not in STOPWORDS]
+        results.append(result)
+    model = Word2Vec.load(model_ser.download())
+    distance = model.wv.wmdistance(*results)
+    logger.info(f"Computing word movers distance for: {SENTENCE_A} and {SENTENCE_B} ")
+    logger.info(f"Word Movers Distance is {distance}")
+
+
+# %%
+# Dimensionality Reduction and Plotting
+# =====================================
+#
+# The word embeddings made by the model can be visualized after reducing the dimensionality to two with t-SNE.
+# This task can take a few minutes to complete.
+@task(cache_version="1.0", cache=True, limits=Resources(mem="1000Mi"))
+def dimensionality_reduction(model_ser: FlyteFile[MODELSER_NLP]) -> plotdata:
+    model = Word2Vec.load(model_ser.download())
+    num_dimensions = 2
+    vectors = np.asarray(model.wv.vectors)
+    labels = np.asarray(model.wv.index_to_key)
+    tsne = TSNE(n_components=num_dimensions, random_state=0)
+    vectors = tsne.fit_transform(vectors)
+    x_vals = [float(v[0]) for v in vectors]
+    y_vals = [float(v[1]) for v in vectors]
+    labels = [str(l) for l in labels]
+    return x_vals, y_vals, labels
+
+
+@task(
+    cache_version="1.0", cache=True, limits=Resources(mem="600Mi"), disable_deck=False
+)
+def plot_with_plotly(x: List[float], y: List[float], labels: List[str]):
+    layout = go.Layout(height=600, width=800)
+    fig = go.Figure(
+        data=go.Scattergl(x=x, y=y, mode="markers", marker=dict(color="aqua")),
+        layout=layout,
+    )
+    indices = list(range(len(labels)))
+    selected_indices = random.sample(indices, 50)
+    for i in selected_indices:
+        fig.add_annotation(
+            text=labels[i],
+            x=x[i],
+            y=y[i],
+            showarrow=False,
+            font=dict(size=15, color="black", family="Sans Serif"),
+        )
+    flytekit.Deck("Word Embeddings", io.to_html(fig, full_html=True))
+
+
+# %%
+# Running the Workflow
+# ====================
+#
+# Let's kick off a workflow! This will return the Word2Vec and LDA models as workflow outputs.
+@workflow
+def nlp_workflow() -> workflow_outputs:
```

Comment:
You can allow users to set hyperparameters by adding them as inputs to the workflow, only if it makes sense.